flocker/node/agents/blockdevice.py

```diff
@@ -1468,6 +1540,94 @@ def _get_system_mounts(self, volumes, compute_instance_id):
+
+        Examine each volume attached to this host.  Determine if the filesystem
+        on the volume has the corresponding dataset_id recorded in its UUID.
+        This is how filesystems are created post-1.0.0 (see FLOC-2369).  If the
+        filesystem does not have this information recorded and the filesystem
+        is not mounted, put it there.  If the filesystem is mounted we cannot
+        modify it.  We will fix it when we notice it is unmounted.
+
+        The intent is for this to cause Flocker 1.0.0-created volumes to look
+        just like they would have if they had been created by Flocker >1.0.0.
+
+        :param IBlockDeviceAPI api: The storage driver responsible for the
+            volumes.
+        :param list volumes: The ``BlockDeviceVolumes`` that exist and will be
+            considered for upgrade.
+        """
+        mounted = {
+            FilePath(partition.device)
+            for partition
+            in psutil.disk_partitions()
+        }
+
+        compute_instance_id = api.compute_instance_id()
+        for volume in _attached_volumes(compute_instance_id, volumes):
+            try:
+                get_device_for_dataset_id(volume.dataset_id)
+                # If we found the dataset_id like this the filesystem is
+                # sufficiently new to not need any changes.  Proceed to the
+                # next volume.
+                continue
+            except KeyError:
+                # There is no filesystem on the volume because volume setup
+                # failed partway through the process or there is no filesystem
+                # with a UUID that matches the dataset id because the volume
+                # was created by 1.0.0.
+                #
+                # The filesystem isn't mounted because the 1.0.1 upgrade
+                # instructions tell the user to restart the node upon upgrade.
+                #
+                # Detach and then re-attach the volume so that get_device_path
+                # can tell us the device.  This is slow and we might be able to
+                # save time by calling get_device_path first.  It might give us
+                # an answer.  Maybe do that optimization.
+                api.detach_volume(volume.blockdevice_id)
+                api.attach_volume(volume.blockdevice_id, compute_instance_id)
+                device_path = api.get_device_path(volume.blockdevice_id)
+
+            # Find out of it is mounted.  Skip it (don't break the rest of the
+            # loop) if it is mounted because there's nothing we can do about
+            # mounted filesystems.  They're not supposed to be mounted because
+            # we told the user to restart the node.
+            if device_path in mounted:
+                FLOCKER_1_0_0_FS_MOUNTED(
+                    block_device_id=volume.blockdevice_id,
+                    block_device_path=device_path,
+                ).write()
```

Comment:
Actions are the eliot abstraction that can succeed or fail. So make sure things that can fail are logged as actions, then raise an exception that finishes an action (e.g. in context manager or result of `eliot.twisted.DeferredContext`-wrapped Deferred).